node/core/provisioner/src/disputes/mod.rs

```diff
@@ -0,0 +1,401 @@
+			total_votes_len += votes_len
+		}
+	}
+
+	result
+}
+
+struct PartitionedDisputes {
+	// Active disputes without enough onchain votes to conclude. These
+	// will be sent to the Runtime with FIRST priority.
+	pub cant_conclude_onchain: Vec<(SessionIndex, CandidateHash)>,
+	// Active disputes completely unknown onchain. Will be sent to the
+	// Runtime with SECOND priority.
+	pub unknown_onchain: Vec<(SessionIndex, CandidateHash)>,
+	// Active disputes with enough votes to conclude onchain. These will
+	// conclude (or have already concluded) even if we don't send any votes.
+	// Anyway - they will be sent to the Runtime with THIRD priority.
+	pub can_conclude_onchain: Vec<(SessionIndex, CandidateHash)>,
+	// Concluded disputes which are known onchain. These are not
+	// interesting and won't be sent to the Runtime.
+	pub concluded_known_onchain: Vec<(SessionIndex, CandidateHash)>,
+	// Concluded local disputes which are completely unknown for the Runtime.
+	// Hopefully this should never happen. Will be sent to the Runtime with FOURTH priority.
+	pub concluded_unknown_onchain: Vec<(SessionIndex, CandidateHash)>,
+}
+
+impl PartitionedDisputes {
+	fn new() -> PartitionedDisputes {
+		PartitionedDisputes {
+			cant_conclude_onchain: Vec::new(),
+			unknown_onchain: Vec::new(),
+			can_conclude_onchain: Vec::new(),
+			concluded_known_onchain: Vec::new(),
+			concluded_unknown_onchain: Vec::new(),
+		}
+	}
+
+	fn into_iter(self) -> impl Iterator<Item = (SessionIndex, CandidateHash)> {
+		self.cant_conclude_onchain
+			.into_iter()
+			.chain(self.unknown_onchain.into_iter())
+			.chain(self.can_conclude_onchain.into_iter())
+			// .chain(self.concluded_known_onchain.into_iter()) -> dropped on purpose
+			.chain(self.concluded_unknown_onchain.into_iter())
+	}
+}
+
+fn partition_recent_disputes(
+	recent: Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	onchain: &HashMap<(u32, CandidateHash), DisputeState>,
+) -> PartitionedDisputes {
+	let mut partitioned = PartitionedDisputes::new();
+
+	// Drop any duplicates
+	let unique_recent = recent
+		.into_iter()
+		.map(|(session_index, candidate_hash, dispute_state)| {
+			((session_index, candidate_hash), dispute_state)
+		})
+		.collect::<HashMap<_, _>>();
+
+	// Split ACTIVE from CONCLUDED disputes
+	let (active, concluded): (
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	) = unique_recent
+		.into_iter()
+		.map(|((session_index, candidate_hash), dispute_state)| {
+			(session_index, candidate_hash, dispute_state)
+		})
+		.partition(|(_, _, status)| *status == DisputeStatus::Active);
+
+	// Split ACTIVE in three groups...
+	for (session_index, candidate_hash, _) in active {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(d) => {
+				// TODO: handle double voting here!!!
+				if d.validators_for.count_ones() + d.validators_against.count_ones() >
+					supermajority_threshold(d.validators_for.len())
```

Comment:
Is this good enough or I'd better get the number of the validators from the session?

Reply:
Bitfields are received from Runtime, so we can trust them at this point. This is good enough.